### DISCUSSION TITLE
Physics with rapier

### DISCUSSION THREAD

**Discussion by aligator** [splashdust/bevy_voxel_world#18] [Q&A]
Hi,

would it be possible to automatically generate colliders for rapier based on the loaded chunks and update whenever a chunk gets updated?
Did you try something like this before?

**Comment by splashdust** [comment]
I have not worked with Rapier before, so I'm not sure how it works exactly, but if you can generate colliders from Bevy meshes somehow, you can get them like this:

```rust
fn generate_chunk_colliders(
  added_chunk_meshes: Query<&Handle<Mesh>, Added<Chunk<MyWorldConfig>>>,
  meshes: Res<Assets<Mesh>>
) {
  for mesh_handle in added_chunk_meshes.iter() {
    if let Some(mesh) = meshes.get(&mesh_handle) {
      // Generate colliders from the mesh
    }
  }
}
```

Then, to react to changed meshes, you could use the [`ChunkWillRemesh`](https://github.com/splashdust/bevy_voxel_world/blob/262b1248bf33e741c71709e147b7d7cd43b93dfb/src/voxel_world.rs#L37) event to get notified when a chunk got remeshed.

Not sure if this will work, but maybe a place to start at least :)

I'd be interested to hear what you come up with, or if you have any suggestions on how to make it easier to integrate physics with `bevy_voxel_world`.

**Reply by aligator** [reply]
https://www.reddit.com/r/bevy/comments/10g719q/generating_a_collider_from_a_mesh/

Looks like this should be possible. I will have to test this😃.

**Reply by aligator** [reply]
I got it working with a first POC.

I was not able to get the mesh directly, but with an extra system it works.
```rust
use crate::map::MainWorld;
use bevy::ecs::world::FilteredEntityRef;
use bevy::prelude::*;
use bevy::utils::info;
use bevy_rapier3d::prelude::*;
use bevy_voxel_world::prelude::{Chunk, ChunkWillRemesh, VoxelWorld};

pub struct PhysicPlugin;
impl Plugin for PhysicPlugin {
    fn build(&self, app: &mut App) {
        app.add_plugins(RapierPhysicsPlugin::<NoUserData>::default())
            .insert_resource(RapierConfiguration {
                gravity: Vec3::new(0.0, -98.0, 0.0),
                ..Default::default()
            })
            .add_systems(
                Update,
                (mark_updated_chunks, generate_chunk_colliders).chain(),
            );
    }
}

#[derive(Component, Default)]
struct NeedsColliderUpdate;

fn mark_updated_chunks(
    mut commands: Commands,
    new_chunk: Query<Entity, Added<Chunk<MainWorld>>>,
    mut event_reader: EventReader<ChunkWillRemesh>,
) {
    // Changed chunks
    for event in event_reader.read() {
        commands.entity(event.entity).insert(NeedsColliderUpdate);
    }

    // New chunks
    for entity in new_chunk.iter() {
        commands.entity(entity).insert(NeedsColliderUpdate);
    }
}

fn generate_chunk_colliders(
    mut commands: Commands,
    updated_chunks: Query<(&Handle<Mesh>, &Chunk<MainWorld>, Entity), With<NeedsColliderUpdate>>,
    meshes: Res<Assets<Mesh>>,
) {
    for (mesh_handle, chunk, entity) in updated_chunks.iter() {
        commands.entity(entity).remove::<NeedsColliderUpdate>();

        let Some(mesh) = meshes.get(mesh_handle) else {
            info!("Failed to get mesh for chunk");
            continue;
        };

        // Generate colliders from the mesh
        if mesh.count_vertices() == 0 {
            info!("Mesh has no vertices");
            continue;
        }
        let Some(collider) = Collider::from_bevy_mesh(mesh, &ComputedColliderShape::TriMesh) else {
            info!("Failed to generate collider for chunk mesh");
            continue;
        };

        info!("Generated collider for chunk mesh {:?}", chunk.position);

        commands.entity(entity).insert((RigidBody::Fixed, collider));
    }
}
```
https://github.com/aligator/bevy-miner

* But as I already suspected, it is a huge performance hit. (note: I already reduced the spawning range)
Maybe this could be minimized by only generating the collider for direct neighbor chunks...
Is there a way to get notified when the player "enters" a new chunk? Then I could listen  to it and rebuild only the near chunks.
* Also it creates A collider for the 'water' cubes. If this should work someday in a usable way, there must be a way to mark some block types as non-colidable (e.g. NonSolid) and generate an additional mesh which exclude them...
* I did not test yet if the rebuilding on block-modifying would be fast enough.

**Reply by splashdust** [reply]
Cool!

It seems like the Rapier debug render has quite a massive impact on frame rate. With that turned off your example runs at 120fps on my machine with a spawn distance of 20.
Reducing to neighbouring chunks may be a good idea regardless though. There is no built-in method to find which chunk the camera is in, but you could try [bevy_mod_raycast](https://github.com/aevyrie/bevy_mod_raycast) and just shoot a ray straight down from the camera to find the related chunk entity you are standing on.

For water you could use a separate voxel world (check [this example](https://github.com/splashdust/bevy_voxel_world/blob/main/examples/multiple_worlds.rs) for how to use multiple worlds) and then only generate colliders for the regular world.

**Reply by aligator** [reply]
Good point ^^ Didn't think that it would have that impact...
I get 60 FPS with a distance of 20 (I only use an internal GPU, so still ok)

Also a good idea with the extra world and raycast, let's see if I can get this working...
Maybe I can also just use the second world to reduce the rendered collider chunks, which would be simpler than using the raycast.

**Reply by aligator** [reply]
Two problems:

* When spawning an extra world just for background processing, is it possible to just disable all rendering of it to save resources?
* how can I listen on a specific world with EventReader<ChunkWillRemesh>

**Reply by splashdust** [reply]
Try inserting a `Visibility::Hidden` component on the chunk entity to hide it.

As for listening to a specific world, I'm afraid the events doesn't support multiple worlds yet. I didn't consider that! They should probably have the Config generic like the other resources, but it's not added yet. If you'd like to take a stab at implementing it, I'm open for pull requests :)

You could also try querying for changed mesh handles (ie `Changed<Handle<Mesh>>` in a query filter). If that works, those events may in fact be redundant.

**Reply by aligator** [reply]
Many thanks for your advices :)

Ok the for the listening I opened a PR. Seems to work just fine.

For the visibility, it seems to interfere with the rapier debug renderer, however I currently add the collider to the physic-world.
Maybe it would be better to add it to the real world. But then I would need a simple way to access the chunk-entity. Currently the voxel world does not expose this. (Or do I miss something here?)

I already experimented with this by exposing `get_chunk` https://github.com/aligator/bevy_voxel_world/blob/e337d74fc59a1f3ea3a3c61c1788f3a127fb31a4/src/voxel_world.rs#L89
And using it like this:
https://github.com/aligator/bevy-miner/blob/28b460011497447adb9673702771f65e387e4df6/src/physic.rs#L44

But with this method it adds some random colliders to (I assume) Point ZERO. Not sure why this happens...

Also I once disabled rendering of the main world, to see if the visibility works.   
-> It flashes for a short time (obviously, because it is hidden shortly after spawning...) 

It would be nice to be able to have some `init_chunk` method in the world configuration, where I can attach custom entities to chunks before they get added to bevy. What do you think?


And also good news: the `Changed<Handle<Mesh>>` method works just fine :)

**Reply by splashdust** [reply]
Yeah, the `get_chunk` method seems like a good thing to have! I guess having access to all the data in `ChunkData` could be useful too, for example if you want to do generate colliders directly from voxel data or do some customized meshing.
It would probably be good to refactor `ChunkData` to not have all the `pub` fields though, and use accessors instead. Either that or add a separate type that `get_chunk` returns...

Not sure what the random collider is about. But it could perhaps be related to the buffered implementation of the chunking system. 

It goes something like this:
1. When a new chunk is spawned, the Bevy entity is first spawned with a `NeedsRemesh` component in the `spawn_chunks` function, and a placeholder is added to the `ChunkMapInsertBuffer`.
2. Then, in the `remesh_dirty_chunks` system, a task is spawned to generate chunk data and mesh in a separate thread.
3. When the task is done, the result is picked up by the `spawn_meshes` function, where the mesh is added to the entity and the updated `ChunkData` is pushed to the `ChunkMapUpdateBuffer`
4. These buffers are then flushed to the actual `chunk_map` in the `flush_chunk_map_buffers` system.

The buffering is there to minimize thread contention by limiting the number of write locks that needs to be acquired on the `chunk_map` RWLock. That way meshing threads don't get held up every time you call `set_voxel`.

I'm thinking that this could lead to a situation where you are able to get the newly spawned or updated chunk entity, but when you call `get_chunk` you get either no result or an outdated result because the buffers has not had a chance to flush yet.

As for the flashing when hiding new chunks, I think that particular issue could be solved by `bevy_voxel_world` spawning all the chunks as children to a world-specific parent entity instead of just putting everything in the top level. Then the parent entity could be set as hidden, and Bevy would not render any chunks in that world. It would also look a lot cleaner when inspecting the Bevy world :)

But that said, an `init_chunk_entity` callback, or similar, would still be useful. I guess it would take an `&EntityCommands` reference as a parameter so that you can `insert()` components. Not sure what the borrow checker will make of that though!

**Reply by aligator** [reply]
Thanks for merging :)

Now I did another experiment:
Instead of using two worlds, I tried to access the voxel data and generate a new mesh without the water blocks.
For this
* I used the get_chunk to retrieve the voxels
* And I exposed the `generate_chunk_mesh` function https://github.com/aligator/bevy_voxel_world/blob/9a2f6c906aedd751db97b133c73309595daa83a6/src/meshing.rs#L27 

Then I could successfully generate the collider mesh https://github.com/aligator/bevy-miner/blob/fe0370354d35a6e7885f08b0034034c2a30cc85c/src/physic.rs#L21

Not sure if the generate_chunk_mesh method should be public like this, also it generates much more than I need for the colliders (e.g. I don't need a texture), but for now it is working with it...

I will open a WIP PR as I am not sure yet what's the best way to expose and use the chunk data and voxels.
